internal/internal_task_pollers.go

```diff
@@ -253,6 +260,90 @@ func (wtp *workflowTaskPoller) processResetStickinessTask(rst *resetStickinessTa
+		wtp.forceRespondDecisionTaskCompleted(wc, workflowTask, startTime)
+	})
+}
+
+func (wtp *workflowTaskPoller) forceRespondDecisionTaskCompleted(wc *WorkflowExecutionContext, workflowTask *workflowTask, startTime time.Time) {
+	wc.Lock()
+	defer wc.release()
+
+	if wc.currentDecisionTask != workflowTask.task {
+		// decision task already completed
+		var currentTaskStartedEventID int64
+		if wc.currentDecisionTask != nil {
+			currentTaskStartedEventID = wc.currentDecisionTask.GetStartedEventId()
+		}
+		wtp.logger.Debug("DecisionTask already completed when force responding timer fires.",
+			zap.Int64("TaskStartedEventID", workflowTask.task.GetStartedEventId()),
+			zap.Int64("CurrentStartedEventID", currentTaskStartedEventID))
+		return
+	}
+
+	completeRequest := wc.CompleteDecisionTaskWithLock(true)
+	wtp.logger.Debug("Force RespondDecisionTaskCompleted.",
+		zap.Int64("TaskStartedEventID", workflowTask.task.GetStartedEventId()))
+	wtp.metricsScope.Counter(metrics.DecisionTaskForceCompleted).Inc(1)
+
+	response, err := wtp.RespondTaskCompletedWithMetrics(completeRequest, workflowTask.task, startTime)
+	if err != nil {
+		return
+	}
+
+	wc.processCompleteDecisionResponseWithLock(response, wtp)
+	return
+}
+
+func (w *WorkflowExecutionContext) processCompleteDecisionResponseWithLock(response *s.RespondDecisionTaskCompletedResponse, wtp *workflowTaskPoller) error {
+	if response == nil || response.DecisionTask == nil {
+		return nil
+	}
+
+	startTime := time.Now()
+	newTask := wtp.toWorkflowTask(response.DecisionTask)
+
+	if newTask.task.History.Events[0].GetEventId() != w.previousStartedEventID+1 {
+		// sanity check for new task, this should not happen.
+		wtp.logger.Info("New task from RespondDecisionTaskCompleted has unexpected events",
+			zap.String(tagWorkflowID, newTask.task.WorkflowExecution.GetWorkflowId()),
+			zap.String(tagRunID, newTask.task.WorkflowExecution.GetRunId()),
+			zap.Int64("CachedPreviousStartedEventID", w.previousStartedEventID),
+			zap.Int64("TaskFirstEventID", newTask.task.History.Events[0].GetEventId()),
+			zap.Int64("TaskStartedEventID", newTask.task.GetStartedEventId()),
+			zap.Int64("TaskPreviousStartedEventID", newTask.task.GetPreviousStartedEventId()))
+
+		w.destroyCachedState()
+		w.resetStateIfDestroyed(newTask.task, newTask.historyIterator)
+	}
+
+	w.setCurrentTask(newTask.task)
+
+	// process new task
+	completedRequest, err := w.processWorkflowTaskWithLock(newTask, true)
+	if err != nil {
+		wtp.logger.Info("Failed to process new task from response of RespondDecisionTaskCompleted.",
+			zap.Error(err))
+		return err
+	}
+
+	if completedRequest == nil {
+		// decision task cannot complete, we need a timer
+		wtp.scheduleRespondDecisionTaskCompleted(w, newTask, startTime)
+		return nil
+	}
+
+	response, err = wtp.RespondTaskCompletedWithMetrics(completedRequest, newTask.task, startTime)
+	if err != nil {
+		return err
+	}
+
+	return w.processCompleteDecisionResponseWithLock(response, wtp)
```

Comment:
Do we have any limit on number of such recursive calls? Could it be that we run out of stack?

Reply:
It only do recursive call when there is new events received on server while worker try to complete decision task. Go stack size limit is 1GB, I think it would be safe here.